### ISSUE TITLE
performance gap between 2d and 3d data

### ISSUE THREAD

**Issue by 674106399** [Jeff-sjtu/res-loglikelihood-regression#9]
Hi,
I'm training by mixing 2d and 3d data, but the 3d-prediction performance is much worse than 2d.
the training log is like this:
![POPO20210817-103512](https://user-images.githubusercontent.com/13503330/129655357-8ba67417-a1c8-4603-94fe-a30dd4e3e51d.png)

the pink line is an Integral Pose model, and the green line is RLE.
As you can see, the 2d performance is competitive(RLE is better in fact), but the 3d performance is worse
Due to I'm doing mix-data training, I calculate my RLE loss as follow:
```
        is_2d = ~is_3d  # masks of 2d and 3d
        num_3d = is_3d.sum()
        num_2d = is_2d.sum()

        t_cnt = 0
        ### joint loss ###
        jc_loss = 0.
        if num_3d > 0:
            jc_loss += self.joint_loss(joint_coord3d[is_3d], nf_loss[is_3d], sigma[is_3d], target_coord[is_3d])
            t_cnt += 1.
        if num_2d > 0:
            jc_loss += self.joint_loss(joint_coord3d[is_2d, :, :2], nf_loss[is_2d, :, :2], sigma[is_2d, :, :2], target_coord[is_2d, :, :2])
            t_cnt += 1.
        jc_loss /= t_cnt
```
Do you have any suggestion?

**Comment by Jeff-sjtu** [comment]
Hi @674106399,

Did you follow our implementation on Human3.6M for mix-data training? If so, it seems that the current implementation has some bugs and I am trying to fix them.

**Comment by 674106399** [comment]
Hi @Jeff-sjtu ,
I think my implementation is similar to yours, but not totally the same. I don't have a 'target_uvd_weight' to mask each coord value, but a 'is_3d' to mask each keypoint. This is why I calculate 2d and 3d losses individually. 
Now I'm trying to modify my code to add target_uvd_weight and do experiment again.
If there's any other thing need to pay attention in mix-data training, please feel free to tell me. thx~

**Comment by jeffffffli** [comment]
Hi @674106399,

We accidentally subtract the `z` value of the root joint during training (now we fix [this line](https://github.com/Jeff-sjtu/res-loglikelihood-regression/blob/master/rlepose/models/regression_nf_3d.py#L153)), which produces many zeros in `bar_mu_3d` and affect the learning of the normalizing flow model. Maybe you can check your implementation to see whether you subtract the root value too.
